Allow user to specify initial guess in AmgXSolver

wcdawn opened the pull request:
I'm using an `AmgXSolver` in a power iteration eigensolver algorithm so the linear solver is called many times. Eventually, I have a pretty good initial guess for the solution vector. I can cut my runtimes in half by specifying an initial guess instead of a zero guess.
<!--GHEX{"id":2132,"author":"wcdawn","editor":"tzanio","reviewers":["artv3","barker29"],"assignment":"2021-03-25T07:34:23-07:00","approval":"2021-03-25T17:55:31.260Z","merge":"2021-03-27T19:11:12.432Z"}XEHG--><!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2132](https://github.com/mfem/mfem/pull/2132) | @wcdawn | @tzanio | @artv3 + @barker29 | 03/25/21 | 03/25/21 | 03/27/21 | |
<!--ELBATXEHG-->

artv3 reviewed (approved):
Looks good to me, but I wonder if false should be the default?

wcdawn commented:
I set it this way to not change the default behavior of the code but I agree that setting `zero_guess=false` by default would probably be better. `zero_guess=false` would reflect the default behavior of, for example, `CGSolver`.

tzanio commented:
It really depends if we expect that its typical use is as a solver or as a preconditioner. In the former case, `zero_guess` should be `false`. In the latter case, it should be `true`.

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

barker29 commented on linalg/amgxsolver.hpp:
```diff
@@ -313,6 +318,8 @@ class AmgXSolver : public Solver
 
    std::string mpi_gpu_mode;
+
+   bool zero_guess = true;
```
Following @tzanio 's comment, can you set the `zero_guess` value based on `amgxMode` in `DefaultParameters()` method so it's `false` for a solver and `true` for a preconditioner?

barker29 reviewed (changes requested):
Other than my minor suggestion, this PR is a nice improvement.

wcdawn commented:
I've changed how this is implemented to avoid having to choose a default and be more consistent with the rest of MFEM. Now, the behavior is based on the `iterative_mode` boolean from the parent `Solver` class. 

I've also implemented @barker29 request to set the iterative mode based on whether the class is being used as a solver or preconditioner. As such, I've removed the setter function because it should be unnecessary now.

Thanks for the help!

barker29 reviewed (approved):
Looks good!

tzanio commented:
@artv3 can you confirm you are OK with the latest changes?

artv3 commented:
Yup looks good to me!

tzanio commented:
Merged in `next` for testing...

tzanio reviewed (approved):
Thanks @wcdawn !
